### ISSUE TITLE
Combining Generators with Promises Using Asynquence

### ISSUE THREAD

**Issue by Aseelaldallal** [getify/asynquence#106]
I'm trying to use a generator to perform an asynchronous operation (readdir) and print the result. I'm trying to use what I learned from your You dont know JS series on Lynda.com .

Here is my code, without promises:

```
var dirname = //some directory name;
function *gen(done) {
    let exercises = yield fs.readdir(dirname, done);
    console.log(exercises);
}

let G = gen(function(err, res) {
    G.next(res);
})

G.next();
```
This works and prints [ 'README.md', 'sol.js' ]

From what I learned, for asynchronous operations, combining generators with promises is better (trust issues). Is this true, or did I misunderstand?

Thus, I tried to combine promises with generators. I used your Asynquence library.

```
const fs = require('fs');
const ASQ = require('asynquence-contrib');

function readDirectory(dir) {
    return ASQ(function(done){
        fs.readdir(dir,done);
    });
}
function *genProm() {
    let exercises = yield readDirectory(dirname);
    console.log(exercises);
}

ASQ().runner(genProm);
```
Now, the result is [ null, [ 'README.md', 'sol.js' ] ]

My Questions are:

Why is the result different here? Where is the null coming from? And how do I get rid of it?
Does it make sense to combine promises and generators here?

**Comment by getify** [comment]
Do you see this part in your first snippet?

```js
function(err, res) {
    G.next(res);
}
```

What's happening here is that a function callback is in the style of "error-first callback" -- that's what `fs.readdir(..)` is expecting to receive for its callback -- meaning the first param is reserved for the error signal. In this form, you're basically ignoring/throwing away any error that might come through, and assuming success.

What'd you'd probably want to do is:

```js
function(err, res) {
    if (err) G.throw(err);
    else G.next(res);
}
```

Here we're splitting the behavior on error vs success, and passing along the appropriate signal (message or exception) to the generator code to handle. This is more robust.

Anyway, either in your version of the snippet, or in mine, you'll notice that the signature maps from 2 inputs (`err,res`) down to one input (`err` or `res`). That's what you're missing in your second snippet.

You pass along the `done` callback to `fs.readdir(..)`, so of course it's going to receive both `err` ***and*** `res`, and that's why you're seeing the `null` in the first argument.

To fix this, you have two options.

Option 1:

```js
function readDirectory(dir) {
    return ASQ(function(done){
        fs.readdir(dir,function(err,...success){
          if (err) done.fail(err);
          else done(...success);
        });
    });
}
```

See how we're now splitting the behavior and either passing along only the error, or only whatever success messages come through?

But, instead of doing *this* manually, asynquence provides a convenience helper that does this automatically.

Option 2:

```js
function readDirectory(dir) {
    return ASQ(function(done){
        fs.readdir(dir,done.errfcb);
    });
}
```

Notice `done.errfcb`. There's no `(..)` because we're not calling that function... we're passing along the error-first version of the `done(..)` callback for `fs.readdir(..)` to call. `done.errfcb` is automatically created for you to use in cases where you need an error-first style callback form. :)

Actually, there's an alternate form for option 2 if you prefer.

Option 2b:

```js
function readDirectory(dir) {
    var sq = ASQ();
    fs.readdir( dir, sq.errfcb() );
    return sq;
}
```

The difference here is that we're constructing the sequence instance (`sq`), then actually calling `errfcb()` on the instance to produce the error-first style callback as a sequence step in `sq`, then passing that callback into `fs.readdir(..)`. Finally, we return that `sq` sequence so someone else can wait on that step to complete (or error)`.

----

A separate issue is also that you should be "listening" for any of those signaled exceptions. You can either do that on the sequence instance directly:

```js
ASQ().runner(genProm)
.or(function(err){ .. });
```

Or, more preferably, you can handle it inside the generator, with a `try..catch`:

```js
function *genProm() {
    try {
        let exercises = yield readDirectory(dirname);
        console.log(exercises);
    }
    catch(err) {
        // ..
    }
}
```

Hope that's helpful info for you!

**Comment by Aseelaldallal** [comment]
Thank you Kyle for taking the time to respond! It means a lot, and I got it! :)
